onnxruntime/core/optimizer/double_qdq_pairs_remover.cc

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/optimizer/double_qdq_pairs_remover.h"
+
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+
+Status DoubleQDQPairsRemover::ApplyImpl(
+    Graph& graph,
+    bool& modified,
+    int /*graph_level*/,
+    const logging::Logger& /*logger*/) const {
+  const GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (const auto& self_index : node_topology_list) {
+    NodeIndex parent_index = 0;
+    NodeIndex child_index = 0;
+    NodeIndex grandchild_index = 0;
+    if (IsNodeRemovable(graph, self_index, parent_index, child_index, grandchild_index)) {
+      graph.RemoveEdge(self_index, child_index, 0, 0);
+      graph.RemoveEdge(child_index, grandchild_index, 0, 0);
+      graph_utils::ReplaceDownstreamNodeInput(graph, *graph.GetNode(grandchild_index), 0, *graph.GetNode(self_index), 0);
+      graph.RemoveNode(child_index);
+      graph.RemoveNode(grandchild_index);
+
+      modified = true;
+    }
+  }
+  return Status::OK();
+}
+
+bool DoubleQDQPairsRemover::IsNodeRemovable(
```

Comment:
>IsNodeRemovable [](http://example.com/codeflow?start=28&length=15)

We need to make sure each pair QDQ has same scale and zeropoint:

[https://github.com/microsoft/onnxruntime/blob/0c6037b5abe571fc43a55ef7a9d2f846820fbe5d/onnxruntime/core/optimizer/qdq_transformer/qdq_util.h#L36](https://github.com/microsoft/onnxruntime/blob/0c6037b5abe571fc43a55ef7a9d2f846820fbe5d/onnxruntime/core/optimizer/qdq_transformer/qdq_util.h#L36)

Reply:
And need to check each node doesn't produce graph output because the output value may change.